File: strategy/entry_b_hedge.py
"""Entry B's weekly hedge: diversify away from Entry A's game without
sacrificing too much safety.

This is the lightweight, sequential companion to ``joint_optimizer.py``: it
treats Entry A's pick (however it was decided -- e.g. ``entry_a_value.py``)
as already fixed, then finds Entry B's best team that

  * isn't from the same game as Entry A's pick, so one game's outcome can
    never eliminate both entries, and
  * clears a minimum win-probability floor (default 65%), so Entry B never
    chases diversification into an unsafe underdog.

Among the survivors it just takes the highest win probability -- no future
value weighting here, since Entry B's whole point in this module is
this-week safety, not season-long value banking.

For a true *joint* optimization that picks both entries' teams together
instead of assuming Entry A's pick is fixed first, see joint_optimizer.py.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Tuple

from data.models import Game
from models.win_prob import resolve_team_win_probability
from state.entries_store import load_used_teams_for_entry

ENTRY_NAME = "Entry B"

# 0-100 scale, matching win_pct everywhere else in the project.
DEFAULT_MIN_WIN_PROB_FLOOR = 65.0


def meets_win_prob_floor(win_pct: Optional[float], floor: float) -> bool:
    return win_pct is not None and win_pct >= floor


@dataclass
class HedgeCandidate:
    team_abbreviation: str
    opponent_abbreviation: Optional[str]
    event_id: Optional[str]
    win_pct: Optional[float]  # 0-100, may be spread-estimated
    win_pct_source: str  # "api" | "spread_estimate" | "unknown"
    spread_detail: Optional[str]


@dataclass
class EntryBHedgeRecommendation:
    week: int
    pick: Optional[HedgeCandidate]
    reasoning: str
    alternatives: List[HedgeCandidate] = field(default_factory=list)
    floor_relaxed: bool = False


def _event_id_for_team(games: List[Game], team_abbreviation: str) -> Optional[str]:
    for game in games:
        if game.home.abbreviation == team_abbreviation or game.away.abbreviation == team_abbreviation:
            return game.event_id
    return None


def _build_candidates(
    current_week_games: List[Game], used_teams: List[str], exclude_event_id: Optional[str]
) -> List[HedgeCandidate]:
    candidates: List[HedgeCandidate] = []
    for game in current_week_games:
        if game.state and game.state != "pre":
            continue
        if exclude_event_id is not None and game.event_id == exclude_event_id:
            continue  # Entry A's game -- picking either side here would oppose Entry A
        spread_detail = game.odds.details if game.odds else None
        for team, opponent, is_home in ((game.home, game.away, True), (game.away, game.home, False)):
            if not team.abbreviation or team.abbreviation in used_teams:
                continue
            resolved = resolve_team_win_probability(game, is_home)
            candidates.append(
                HedgeCandidate(
                    team_abbreviation=team.abbreviation,
                    opponent_abbreviation=opponent.abbreviation,
                    event_id=game.event_id,
                    win_pct=resolved.win_pct,
                    win_pct_source=resolved.source,
                    spread_detail=spread_detail,
                )
            )
    return candidates


def rank_hedge_candidates(
    current_week_games: List[Game],
    used_teams: List[str],
    exclude_event_id: Optional[str] = None,
    min_win_prob_floor: float = DEFAULT_MIN_WIN_PROB_FLOOR,
) -> Tuple[List[HedgeCandidate], bool]:
    """Rank eligible candidates by win probability, highest first.

    Returns ``(candidates, floor_relaxed)``. ``floor_relaxed`` is True only
    when at least one candidate existed but none cleared the floor, so we
    fell back to the unfiltered ranking rather than leaving Entry B without
    a pick.
    """
    all_candidates = _build_candidates(current_week_games, used_teams, exclude_event_id)
    all_candidates.sort(key=lambda c: (c.win_pct is None, -(c.win_pct or 0)))

    above_floor = [c for c in all_candidates if meets_win_prob_floor(c.win_pct, min_win_prob_floor)]
    if above_floor:
        return above_floor, False
    return all_candidates, bool(all_candidates)


def _describe(candidate: HedgeCandidate) -> str:
    win_pct = f"{candidate.win_pct:.1f}%" if candidate.win_pct is not None else "unknown"
    basis = " (estimated from spread)" if candidate.win_pct_source == "spread_estimate" else ""
    spread = f", spread {candidate.spread_detail}" if candidate.spread_detail else ""
    return f"{candidate.team_abbreviation} vs {candidate.opponent_abbreviation or '?'} -- {win_pct} win prob{basis}{spread}"


def recommend(
    current_week_games: List[Game],
    current_week: int,
    used_teams: Optional[List[str]] = None,
    entry_a_pick_team: Optional[str] = None,
    min_win_prob_floor: float = DEFAULT_MIN_WIN_PROB_FLOOR,
) -> EntryBHedgeRecommendation:
    """Entry B's top pick for the week, hedged against Entry A's game.

    ``used_teams`` defaults to loading ``state/used_teams_b.json``.
    ``entry_a_pick_team`` is Entry A's team abbreviation for the week (from
    however you decided it, e.g. ``entry_a_value.recommend()``); pass
    ``None`` to rank Entry B without any hedging constraint.
    """
    if used_teams is None:
        used_teams = load_used_teams_for_entry(ENTRY_NAME)

    exclude_event_id = (
        _event_id_for_team(current_week_games, entry_a_pick_team) if entry_a_pick_team else None
    )

    ranked, floor_relaxed = rank_hedge_candidates(
        current_week_games, used_teams, exclude_event_id, min_win_prob_floor
    )

    if not ranked:
        return EntryBHedgeRecommendation(
            week=current_week,
            pick=None,
            reasoning="No eligible teams available this week (all used, or no game data).",
        )

    top, alternatives = ranked[0], ranked[1:]

    parts = [f"Top pick: {_describe(top)}."]
    if floor_relaxed:
        parts.append(
            f"No available team cleared the {min_win_prob_floor:.0f}% floor this week; "
            f"used the safest option available rather than leave Entry B without a pick."
        )
    else:
        parts.append(f"Clears the {min_win_prob_floor:.0f}% win-probability floor.")
    if exclude_event_id is not None:
        parts.append(f"Avoided Entry A's game ({entry_a_pick_team}) entirely, so one result can't eliminate both entries.")
    if alternatives:
        parts.append(f"Next best was {_describe(alternatives[0])}.")

    return EntryBHedgeRecommendation(
        week=current_week,
        pick=top,
        reasoning=" ".join(parts),
        alternatives=alternatives,
        floor_relaxed=floor_relaxed,
    )

File: deadpool/src/engine/strategies/entry-b-hedge.js
/**
 * Entry B's hedge: diversify away from Entry A's game without getting reckless.
 *
 * Port of `strategy/entry_b_hedge.py`, unchanged.
 *
 * The lightweight, sequential companion to the joint optimizer. It treats
 * Entry A's pick as already fixed — however it was decided — then finds Entry
 * B's best team that
 *
 *   * is not from the same game as Entry A's pick, so one result can never
 *     eliminate both entries, and
 *   * clears a minimum win-probability floor, so B never chases
 *     diversification into an unsafe underdog.
 *
 * Among the survivors it simply takes the highest probability. There is no
 * future-value weighting here on purpose: B's whole job in this strategy is
 * this-week safety, not season-long value banking.
 *
 * The floor is relaxed rather than enforced when nothing clears it, and that
 * is called out in the reasoning. A week with no pick is worse than a week
 * with a nervous one, and a floor that silently produced nothing would be
 * discovered at 12:55 on a Sunday.
 */

import { buildOptions, notUsed, byWinPctDesc } from '../constraints.js';
import { f0, f1 } from '../fmt.js';
import { basisPhrase } from '../win-prob.js';
import { recommend as valueRecommend } from './entry-a-value.js';

export const ID = 'hedge';

// 0–100, matching winPct everywhere else.
export const DEFAULT_MIN_WIN_PROB_FLOOR = 65.0;

export const meetsWinProbFloor = (winPct, floor) =>
  winPct !== null && winPct !== undefined && winPct >= floor;

/** The event a team is playing in this week, or null. */
export function eventIdForTeam(games, teamAbbreviation) {
  for (const game of games) {
    if (game.home.abbreviation === teamAbbreviation || game.away.abbreviation === teamAbbreviation) {
      return game.eventId;
    }
  }
  return null;
}

function buildCandidates(games, usedTeams, excludeEventId) {
  const options = buildOptions(games).filter(
    // Entry A's game: picking either side of it would put both entries on one
    // result, which is the whole thing this strategy exists to avoid.
    (o) => !(excludeEventId !== null && excludeEventId !== undefined && o.eventId === excludeEventId),
  );
  return notUsed(options, usedTeams).map((o) => ({
    teamAbbreviation: o.teamAbbreviation,
    teamName: o.teamName,
    opponentAbbreviation: o.opponentAbbreviation,
    eventId: o.eventId,
    startDate: o.startDate,
    winPct: o.winPct,
    winPctSource: o.winPctSource,
    winPctIsEstimated: o.winPctSource === 'spread_estimate',
    spreadDetail: o.spreadDetail,
  }));
}

/**
 * Eligible candidates, best first, and whether the floor had to be relaxed.
 *
 * `floorRelaxed` is true only when candidates existed but none cleared —
 * an empty board is not a relaxed floor, it is an empty board.
 */
export function rankHedgeCandidates(games, usedTeams, excludeEventId = null, minWinProbFloor = DEFAULT_MIN_WIN_PROB_FLOOR) {
  const all = buildCandidates(games, usedTeams, excludeEventId).sort(byWinPctDesc);
  const aboveFloor = all.filter((c) => meetsWinProbFloor(c.winPct, minWinProbFloor));
  if (aboveFloor.length) return { candidates: aboveFloor, floorRelaxed: false };
  return { candidates: all, floorRelaxed: all.length > 0 };
}

export function describe(candidate) {
  const winPct = candidate.winPct === null || candidate.winPct === undefined ? 'unknown' : `${f1(candidate.winPct)}%`;
  const basis = basisPhrase(candidate.winPctSource);
  const spread = candidate.spreadDetail ? `, spread ${candidate.spreadDetail}` : '';
  return `${candidate.teamAbbreviation} vs ${candidate.opponentAbbreviation || '?'} -- ${winPct} win prob${basis}${spread}`;
}

export function recommend(games, currentWeek, usedTeams, entryAPickTeam = null, minWinProbFloor = DEFAULT_MIN_WIN_PROB_FLOOR) {
  const excludeEventId = entryAPickTeam ? eventIdForTeam(games, entryAPickTeam) : null;
  const { candidates, floorRelaxed } = rankHedgeCandidates(games, usedTeams, excludeEventId, minWinProbFloor);

  if (!candidates.length) {
    return {
      week: currentWeek,
      pick: null,
      reasoning: 'No eligible teams available this week (all used, or no game data).',
      alternatives: [],
      floorRelaxed: false,
    };
  }

  const [top, ...alternatives] = candidates;

  const parts = [`Top pick: ${describe(top)}.`];
  if (floorRelaxed) {
    parts.push(
      `No available team cleared the ${f0(minWinProbFloor)}% floor this week; `
      + 'used the safest option available rather than leave Entry B without a pick.',
    );
  } else {
    parts.push(`Clears the ${f0(minWinProbFloor)}% win-probability floor.`);
  }
  if (excludeEventId !== null && excludeEventId !== undefined) {
    parts.push(`Avoided Entry A's game (${entryAPickTeam}) entirely, so one result can't eliminate both entries.`);
  }
  if (alternatives.length) {
    parts.push(`Next best was ${describe(alternatives[0])}.`);
  }

  return { week: currentWeek, pick: top, reasoning: parts.join(' '), alternatives, floorRelaxed };
}

/* ------------------------------------------------ the registry contract -- */

/**
 * Entry A decided on value, Entry B hedged against it.
 *
 * This composition is what `entry_b_hedge.py`'s own docstring describes —
 * "treats Entry A's pick (however it was decided -- e.g. entry_a_value.py) as
 * already fixed" — and it was never wired to anything. Nothing new is computed
 * here: it calls the two ported modules in the order they were written for.
 */
export default {
  id: 'sequential',
  name: 'One safe pick, one hedged',
  blurb: 'Your first entry takes the best pick it can find. Your second takes the safest team '
    + 'from a different game, so no single result can knock both entries out at once.',
  entries: 'both',
  params: [
    { key: 'minWinProbFloorB', label: 'Entry B floor', type: 'percent', default: DEFAULT_MIN_WIN_PROB_FLOOR, min: 0, max: 99, help: 'How safe Entry B has to be before diversification is worth it. Relaxed, loudly, if nothing clears it.' },
    { key: 'lookaheadWeeks', label: 'Entry A looks ahead', type: 'int', default: 6, min: 1, max: 12, unit: 'weeks' },
  ],

  run(ctx) {
    const [a, b] = ctx.entries;
    const floor = ctx.params.minWinProbFloorB ?? DEFAULT_MIN_WIN_PROB_FLOOR;
    const lookahead = ctx.params.lookaheadWeeks ?? 6;

    const aResult = valueRecommend(ctx.games, ctx.schedule, ctx.week, ctx.usedTeams[a.id] ?? [], lookahead);
    const bResult = recommend(
      ctx.games, ctx.week, ctx.usedTeams[b.id] ?? [],
      aResult.pick ? aResult.pick.teamAbbreviation : null, floor,
    );

    const warnings = [];
    if (bResult.floorRelaxed) {
      warnings.push({ level: 'warn', text: `Nothing available to ${b.name} cleared the ${f0(floor)}% floor. The safest option was taken instead of leaving the entry without a pick.` });
    }
    if (aResult.pick && bResult.pick && aResult.pick.eventId === bResult.pick.eventId) {
      warnings.push({ level: 'danger', text: 'Both picks are in the same game — one result would eliminate both entries.' });
    }

    return {
      strategyId: 'sequential',
      picks: [
        { entry: a.id, candidate: aResult.pick, reasoning: aResult.reasoning, factors: aResult.pick ? [
          { label: 'Win probability', value: `${f1(aResult.pick.winPct)}%`, weight: 1, note: aResult.pick.winPctIsEstimated ? 'Estimated from the spread.' : 'From ESPN.' },
          { label: 'Score after lookahead', value: aResult.pick.score === null ? null : f1(aResult.pick.score), weight: 1 },
        ] : [] },
        { entry: b.id, candidate: bResult.pick, reasoning: bResult.reasoning, factors: bResult.pick ? [
          { label: 'Win probability', value: `${f1(bResult.pick.winPct)}%`, weight: 1, note: bResult.pick.winPctIsEstimated ? 'Estimated from the spread.' : 'From ESPN.' },
          { label: 'Floor', value: `${f0(floor)}%`, weight: 0, note: bResult.floorRelaxed ? 'Relaxed — nothing cleared it.' : 'Cleared.' },
          { label: 'Different game from ' + a.name, value: aResult.pick ? 'yes' : 'n/a', weight: 1, note: aResult.pick ? `${a.name} is on ${aResult.pick.teamAbbreviation}.` : `${a.name} has no pick to avoid.` },
        ] : [] },
      ],
      candidates: {
        [a.id]: aResult.pick ? [aResult.pick, ...aResult.alternatives] : [],
        [b.id]: bResult.pick ? [bResult.pick, ...bResult.alternatives] : [],
      },
      considered: (aResult.alternatives.length + 1) + (bResult.alternatives.length + 1),
      warnings,
    };
  },
};
